### ISSUE TITLE
batch import not running

### ISSUE THREAD

**Issue by isolver** [isolver/OpenHandWrite#134]
I can't get the batch report .bat file to run. Running from the script within Spyder, but with I think the right working directory. Sorry, Sol, should have checked before you went away. I get this...

```
runfile('C:/openWrite/OpenHandWrite/markwrite/batchreportgen.py', wdir='C:/openWrite/OpenHandWrite/markwrite')
Traceback (most recent call last):

  File "<ipython-input-1-c7fcfeb026ed>", line 1, in <module>
    runfile('C:/openWrite/OpenHandWrite/markwrite/batchreportgen.py', wdir='C:/openWrite/OpenHandWrite/markwrite')

  File "C:\openWrite\OpenHandWrite\WinPython\python-2.7.10.amd64\lib\site-packages\spyderlib\widgets\externalshell\sitecustomize.py", line 685, in runfile
    execfile(filename, namespace)

  File "C:\openWrite\OpenHandWrite\WinPython\python-2.7.10.amd64\lib\site-packages\spyderlib\widgets\externalshell\sitecustomize.py", line 71, in execfile
    exec(compile(scripttext, filename, 'exec'), glob, loc)

  File "C:/openWrite/OpenHandWrite/markwrite/batchreportgen.py", line 8, in <module>
    import markwrite

  File "markwrite\__init__.py", line 28, in <module>
    usersettings = readPickle(appdirs.user_config_dir,u'usersettings.pkl')

  File "markwrite\file_io.py", line 303, in readPickle
    return cPickle.load(f)

SystemError: ('unable to find to find type: QString', <built-in function _unpickle_type>, ('PyQt4.QtCore', 'QString', (u'Alt+PgUp',)))
```

**Comment by isolver** [comment]
When you run the "Run_BatchReport.bat" file, what error do you get? 

The error posted seems to be from running it within spyder, which is likely a separate issue (which should still be fixed)

**Comment by isolver** [comment]
I fixed the cause of the error listed above (I think). Now QString prefs are saved as python unicodes and QColor prefs are saved as a (r,g,b) tuple. When file is read, conversion from (r,g,b) tuples to QColors are made.

I am not sure if this will be the only issue you run into when trying to use the script in spyder, please let me know if you get others.

**Comment by isolver** [comment]
Note that you will need to delete your existing app preferences file the first time you run the updated code or you will get an error:

C:\Users\Sol\AppData\Local\MarkWrite\MarkWrite*.*

**Comment by isolver** [comment]
Reopenning, getting following error when I run Run_BatchReport.bat

```
Traceback (most recent call last):
  File "batchreportgen.py", line 27, in <module>
    mwp = MarkWriteProject(file_path=absdfp)
  File "D:\Dropbox\OpenHandWriteDistribution\OpenHandWrite\markwrite\markwrite\project.py", line 251, in __init__
    self.createNewProject(file_path, fimporter)
  File "D:\Dropbox\OpenHandWriteDistribution\OpenHandWrite\markwrite\markwrite\project.py", line 589, in createNewProject
    MarkWriteProject._selectedtimeregion.project = self
AttributeError: 'NoneType' object has no attribute 'project'
FreeType import Failed: Freetype library not found
Press any key to continue . . .
```

**Comment by Mark-Torrance** [comment]
Sorry, should have done this before. I get the same when running from the bat file.

```
Updated usersettings.pkl in: C:\Users\mark torrance\AppData\Local\MarkWrite\Mark
Write
txyp mwp False
Traceback (most recent call last):
  File "batchreportgen.py", line 27, in <module>
    mwp = MarkWriteProject(file_path=absdfp)
  File "C:\openWrite\OpenHandWrite\markwrite\markwrite\project.py", line 249, in
 __init__
    self.createNewProject(file_path, fimporter)
  File "C:\openWrite\OpenHandWrite\markwrite\markwrite\project.py", line 539, in
 createNewProject
    MarkWriteProject._selectedtimeregion.project = self
AttributeError: 'NoneType' object has no attribute 'project'
FreeType import Failed: Freetype library not found
Press any key to continue . . .
```

**Comment by isolver** [comment]
Fixed the errors we both reported yesterday, so running Run_BatchReport.bat now works.

I also updated the example script so that it generates a report for all supported files in the test_data folder. I have 2 markwrite project files saved in this folder on my PC, and the batch reporting code creates reports for the saved project files correctly as well. 

@Mark-Torrance : When you get the latest code, please retest from within spyder, ensuring cwd is set to be the parent dir of the 'test_data' folder, and let me know if you get the Qt error still. If it is OK, please close.

**Comment by Mark-Torrance** [comment]
Runs fine from bat file. I get the following running in Spyder.

Tutorial question: It's right, there isn't a module called MarkWrite (unless there's one buried somewhere in the dirs that are added to the path when the .bat runs, but aren't in the path for spyder. I've put the Spyder path below.

```
  File "<ipython-input-1-9dc1ae65dcd6>", line 1, in <module>
    runfile('C:/openWrite/OpenHandWrite/markwrite/batchreportgen.py', wdir='C:/openWrite/OpenHandWrite/markwrite/test_data')

  File "C:\openWrite\OpenHandWrite\WinPython\python-2.7.10.amd64\lib\site-packages\spyderlib\widgets\externalshell\sitecustomize.py", line 685, in runfile
    execfile(filename, namespace)

  File "C:\openWrite\OpenHandWrite\WinPython\python-2.7.10.amd64\lib\site-packages\spyderlib\widgets\externalshell\sitecustomize.py", line 71, in execfile
    exec(compile(scripttext, filename, 'exec'), glob, loc)

  File "C:/openWrite/OpenHandWrite/markwrite/batchreportgen.py", line 7, in <module>
    import markwrite

ImportError: No module named markwrite
```

``` python
sys.path
Out[3]: 
['',
 'C:\\openWrite\\OpenHandWrite\\WinPython\\python-2.7.10.amd64\\lib\\site-packages\\psychopy-1.83.2-py2.7.egg',
 'C:\\openWrite\\OpenHandWrite\\WinPython\\python-2.7.10.amd64\\python27.zip',
 'C:\\openWrite\\OpenHandWrite\\WinPython\\python-2.7.10.amd64\\DLLs',
 'C:\\openWrite\\OpenHandWrite\\WinPython\\python-2.7.10.amd64\\lib',
 'C:\\openWrite\\OpenHandWrite\\WinPython\\python-2.7.10.amd64\\lib\\plat-win',
 'C:\\openWrite\\OpenHandWrite\\WinPython\\python-2.7.10.amd64\\lib\\lib-tk',
 'C:\\openWrite\\OpenHandWrite\\WinPython\\python-2.7.10.amd64',
 'C:\\openWrite\\OpenHandWrite\\WinPython\\python-2.7.10.amd64\\lib\\site-packages',
 'C:\\openWrite\\OpenHandWrite\\WinPython\\python-2.7.10.amd64\\lib\\site-packages\\FontTools',
 'C:\\openWrite\\OpenHandWrite\\WinPython\\python-2.7.10.amd64\\lib\\site-packages\\win32',
 'C:\\openWrite\\OpenHandWrite\\WinPython\\python-2.7.10.amd64\\lib\\site-packages\\win32\\lib',
 'C:\\openWrite\\OpenHandWrite\\WinPython\\python-2.7.10.amd64\\lib\\site-packages\\Pythonwin',
 'C:\\openWrite\\OpenHandWrite\\WinPython\\python-2.7.10.amd64\\lib\\site-packages\\IPython\\extensions',
 'C:\\openWrite\\OpenHandWrite\\WinPython\\settings\\.ipython']
```

**Comment by Mark-Torrance** [comment]
~~Tutorial question: It's right, there isn't a module called MarkWrite (unless there's one buried somewhere in the dirs that are added to the path when the .bat runs, but aren't in the path for spyder.~~ Sorry, I've answered my own question, I think. Still don't understand why I get that error, though.

**Comment by isolver** [comment]
Sorry Mark, I am sure I answered this one as well. Either I am loosing it or a couple of my replies did not get processed by github that evening. Anyhow, sorry again.

>  It's right, there isn't a module called MarkWrite

The folder OpenHandWrite/markwrite/markwrite is a python package, but it is not 'installed', so it is not in the a folder in the default python path. The current working dir is in the python default path, that is why markwrite is found when scripts are run when the cwd is OpenHandWrite/markwrite.

Making a setup.py file for the markwriote package, so it can be installed like standard python packages using 'pip' or 'easy-install', should be on the future to do list. I think I already added an issue to the tracker for this when I wrote the missing reply a couple days ago .

For now, if someone is using the OpenHandWrite python distro, it would be easy to add markwrite to the python path so it is found regardless of the CWD. Should I do that?

> Sorry, I've answered my own question, I think. 

What was your answer? ;)

> Still don't understand why I get that error, though.

It looks like the CWD is being set to "C:/openWrite/OpenHandWrite/markwrite/test_data" by the ipython env. that is running the script and generating the error you give above.

Since C:/...../OpenHandWrite/markwrite/ folder is not the CWD and is not in the python path, ipython / spyder is not finding the markwrite package (C:/openWrite/OpenHandWrite/markwrite/markwrite).

**Comment by isolver** [comment]
@Mark-Torrance: did you get the download link for the self contained markwrite folder with a markwrite.exe? Does that work fine on your PC? 

The self contained markwrite folder is ~ 256 MB uncompressed and zips up into a 56 MB archive file. Not too bad.

**Comment by Mark-Torrance** [comment]
> > Sorry, I've answered my own question, I think.
> 
> What was your answer? ;)

Well, basically what you've said: it's a package not a module, so needs to be explicitly in the path.

When Python can find Markwrite (i.e. by Adding OpenHandWrite\markwrite\markwrite or OpenHandWrite\markwrite\ to the python path and or changing cwd to markwrite), it kills python:

``` python
import markwrite
It seems the kernel died unexpectedly. Use 'Restart kernel' to continue using this console.
```

Sorry I'm being slow here. But I am learning!

**Comment by isolver** [comment]
This maybe some issue that is being hidden by the ipython kernel.

Can you try:

1) Open a WinPython Command Prompt. Use "OpenHandWrite/WinPython/WinPython Command Prompt.exe".
2) In the command prompt type:

```
cd ..\..\markwrite
python
```

3) Then in the python interpreter you started in 2):

```
import markwrite
print markwrite.__version__
```

Does this work, or give an error?

**Comment by isolver** [comment]
BTW, I am rearranging the OpenHandWrite project, the important point is that markwrite package will now be installed as a python package.
1. There is now a src\markwrite folder that contains the markwrite package source, including a setup.py.
2. When the OpenHandWrite distribution get built, both psychopy and markwrite get installed into site-packages. So markwrite package is now in python path.
3. The 'MarkWrite" folder within OpenHandWrite distribution root folder now contains just the files needed to run MarkWrite app, not the full source. So this dir now includes only:

```
 Directory of D:\Dropbox\OpenHandWriteDistribution\OpenHandWrite\MarkWrite

30/01/2016  07:00 PM    <DIR>          .
30/01/2016  07:00 PM    <DIR>          ..
26/01/2016  08:09 AM             2,324 batchreportgen.py
17/12/2015  01:55 PM             1,533 customreports.py
30/01/2016  06:32 PM             1,658 runapp.py
30/01/2016  07:00 PM    <DIR>          test_data
               5 File(s)        103,884 bytes
```

In other words, I'm cleaning this all up properly. ;)

**Comment by isolver** [comment]
I have finished the rearrangement and tested things. All seems to work for me.  You should now be able to import markwrite package like you would a standard python package in your custom scripts. 

I'm making a new OpenHandWrite distribution. Will send out URL for download when it is available.

**Comment by Mark-Torrance** [comment]
Using the new distro (310116), and running python from the winpython command prompt (i.e. not via ipython) gives...

```
>>> HDF5_DISABLE_VERSION_CHECK = 2
>>> import markwrite
Warning! ***HDF5 library version mismatched error***
The HDF5 header files used to compile this application do not match
the version used by the HDF5 library to which this application is linked.
Data corruption or segmentation faults may occur if the application continues.

[snip]

This application has requested the Runtime to terminate it in an unusual way.
Please contact the application's support team for more information.
```

**Comment by Mark-Torrance** [comment]
New distro, and running from bat files:
- MarkWrite runs fine
- BatchReport runs fine

**Comment by isolver** [comment]
For now, that error can be fixed by setting the HDF5_DISABLE_VERSION_CHECK env. Variable in the cmd prompt before starting Python:

So at the winpython cmd prompt:

```
set  HDF5_DISABLE_VERSION_CHECK=2
python
```

Then in the python interpreter just started:

```
import markwrite

```

---

Edit: Btw, I'll fix the issue that results in python exiting if hdf5 lib is used without setting HDF5_DISABLE_VERSION_CHECK=2  env. variable first. I must have updated a package that included a newer version of the hdf5.dll than what the pytables package was built with.

**Comment by Mark-Torrance** [comment]
Ah. That works (from the command prompt). I actually thought I'd already tried it, but clearly got something wrong.

**Comment by Mark-Torrance** [comment]
But still issues with ipython, even after adding set HDF5_DISABLE_VERSION_CHECK=2 to the winpython env bat. Closest I've got is this:

```
os.environ["HDF5_DISABLE_VERSION_CHECK"] = "2"

import markwrite
Traceback (most recent call last):

  File "<ipython-input-7-c9d7fa9925c3>", line 1, in <module>
    import markwrite

SystemError: NULL result without error in PyObject_Call
```

So at least no crash now.

**Comment by isolver** [comment]
How are you running the above in ipython? Is it via the Spyder ipython prompt tab or by running ipython Qt Console, notebook, or other?

I just want to be sure I am testing the same way you are running. Thanks.

**Comment by isolver** [comment]
Disregard my last post; I see it happening when I use the Spyder ipython console tab. 

Not sure if it is related or not, but the following is printed to the Spyder 'Console' tab after you do the markwrite import in ipython:

```
exception UnsupportedOperation('IOStream has no fileno.',) while fixing up sys.stdout and sys.stderr
```

So the latest ipython issue may have nothing to do with hdf5

**Comment by isolver** [comment]
Other observations when testing "import markwrite" within a python file being edited and launching using the WinPython/Spyder.exe:
- **WORKS**: Run the file with the Spyder Run Settings Console Option set to  **"Execute in an external System terminal"** 
- **FAILS**:Run the file with the Spyder Run Settings Console Option set to  **"Execute in current Python or IPython console"**:

```
Python 2.7.10 (default, May 23 2015, 09:44:00) [MSC v.1500 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> runfile('D:/Dropbox/OpenHandWriteDistribution/OpenHandWrite/WinPython/settings/.spyder2/temp.py', wdir='D:/Dropbox/OpenHandWriteDistribution/OpenHandWrite/WinPython/settings/.spyder2')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "D:\Dropbox\OpenHandWriteDistribution\OpenHandWrite\WinPython\python-2.7.10.amd64\lib\site-packages\spyderlib\widgets\externalshell\sitecustomize.py", line 685, in runfile
    execfile(filename, namespace)
  File "D:\Dropbox\OpenHandWriteDistribution\OpenHandWrite\WinPython\python-2.7.10.amd64\lib\site-packages\spyderlib\widgets\externalshell\sitecustomize.py", line 71, in execfile
    exec(compile(scripttext, filename, 'exec'), glob, loc)
  File "D:/Dropbox/OpenHandWriteDistribution/OpenHandWrite/WinPython/settings/.spyder2/temp.py", line 7, in <module>
    import markwrite
SystemError: NULL result without error in PyObject_Call
```
-  **FAILS**: Run the file with the Spyder Run Settings Console Option set to  **"Execute in new dedicated Python console"**:

```
Traceback (most recent call last):
  File "D:\Dropbox\OpenHandWriteDistribution\OpenHandWrite\WinPython\settings\.spyder2\temp.py", line 7, in <module>
    import markwrite
  File "D:\Dropbox\OpenHandWriteDistribution\OpenHandWrite\WinPython\python-2.7.10.amd64\lib\site-packages\markwrite-0.2.2.1-py2.7.egg\markwrite\__init__.py", line 36, in <module>
    writePickle(appdirs.user_config_dir,u'usersettings.pkl', SETTINGS)
  File "D:\Dropbox\OpenHandWriteDistribution\OpenHandWrite\WinPython\python-2.7.10.amd64\lib\site-packages\markwrite-0.2.2.1-py2.7.egg\markwrite\file_io.py", line 362, in writePickle
    if isinstance(v,QtCore.QString):
AttributeError: 'module' object has no attribute 'QString'
```

So I think we should separate issues with running the batch import script in general, vs. issues with importing markwrite when using Spyder / ipython.

If you agree, I will close this bug then, since it works when being run directly using python env. provided, but is failing when run in Spyder via an ipython environment.

**Comment by Mark-Torrance** [comment]
Yes to starting new issue. Sorry that this is being a pain. I think it's worth fixing, though. It's the one part of the project that is going to involve "scientist programming" - the sort of iterative try-it-and-see stuff that semi-competent python users like myself go in for - and so being able to use an IDE is important. ipython less so, I suppose.

**Comment by isolver** [comment]
Agreed. Having Spyder included with WinPython, which is included with the
OpenHandWrite distro, is very important for people who do want to do some
custom python development for MarkWrite or GetWrite stuff and do not want
to have to deal with setting up a Python interpreter and all the
dependencies needed by psychopy, iohub, and markwrite.

Having tools like Spyder included in the OpenHandWrite distro also
justifies the large disk space footprint it takes up.

**Comment by isolver** [comment]
Closing this issue; openned issue #152.

**Comment by Mark-Torrance** [comment]
If you ignore the HDF5 error then, yes, this is Spyder-specific. ipython console, launched directly from IPython Qt Console.exe works...
```python
import os

os.environ["HDF5_DISABLE_VERSION_CHECK"] = "2". 

import markwrite

markwrite
Out[5]: <module 'markwrite' from 'C:\openWrite\OpenHandWrite\WinPython\python-2.7.10.amd64\lib\site-packages\markwrite-0.2.2.1-py2.7.egg\markwrite\__init__.pyc'>
```

but both ipython and non-ipython consoles in Spyder don't find markwrite (or something). This is oridinary python interpreter in Spyder:

```Python
Python 2.7.10 (default, May 23 2015, 09:44:00) [MSC v.1500 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.environ["HDF5_DISABLE_VERSION_CHECK"] = "2"
>>> import markwrite
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
SystemError: NULL result without error in PyObject_Call
```
